toolchain/lower/file_context.cpp

```diff
@@ -190,49 +190,31 @@ auto FileContext::GetOrCreateFunction(SemIR::FunctionId function_id,
   return result;
 }
 
-auto FileContext::BuildFunctionDecl(SemIR::FunctionId function_id,
-                                    SemIR::SpecificId specific_id)
-    -> llvm::Function* {
-  const auto& function = sem_ir().functions().Get(function_id);
-
-  // Don't lower generic functions. Note that associated functions in interfaces
-  // have `Self` in scope, so are implicitly generic functions.
-  if (function.generic_id.has_value() && !specific_id.has_value()) {
-    return nullptr;
-  }
-
-  // Don't lower builtins.
-  if (function.builtin_function_kind != SemIR::BuiltinFunctionKind::None) {
-    return nullptr;
+auto FileContext::BuildFunctionTypeInfo(
+    const SemIR::Function& function, SemIR::SpecificId specific_id,
+    llvm::ArrayRef<SemIR::InstId> implicit_param_patterns,
+    llvm::ArrayRef<SemIR::InstId> param_patterns,
+    const SemIR::ReturnTypeInfo& return_info,
+    llvm::SmallVector<llvm::Type*>& param_types,
+    llvm::SmallVector<SemIR::InstId>& param_inst_ids, llvm::Type*& return_type,
+    SemIR::InstId& return_param_id) -> void {
+  return_type = nullptr;
+  if (return_info.type_id.has_value()) {
+    if (!types_[return_info.type_id.index]) {
```

Comment:
The old code was checking if `type_id.has_value()`. Using it as an index when it doesn't have a value seems incorrect - how come we're not checking the same condition anymore?

Reply:
Ah, seems the checking type and non-null types_ entry are equivalent to testing `return_info.is_valid()`, so I switched to that.

So the code now looks closer to the original - with the difference that the CHECK is now a conditional to return the trivial type.